rfcs/20220404-troubleshooting-helpers/README.md

```diff
@@ -0,0 +1,304 @@
+However, this is trickier for the C API. In that case, if we break
+down a status into multiple statuses (e.g. instead of returning
+`invalid_argument`, we return `invalid_argument_X`), then it becomes
+hardly programmable if a user wants to catch all those.
+
+
+Another aspect is that it does not allow to help our user for
+non-failing diagnostics. For example, dispatching information cannot
+be obtained from status, as:
+- the primitive descriptor was created successfully, even though the
+  dispatched implementation is not the one expected by the user.
+- each implementation that was not dispatched might have a different
+  reason for not dispatching
+
+For the reasons above, we propose to leave the statuses as they
+currently are, so that they align with std::exception, and can be
+easily mapped from C to C++ API.
+
+
+## 2.2 Release mode verbose vs debug mode verbose
+Some debugging information might impact the performance of the
+library, in particular primitive creation and execution time.
+Currently, we already have such debugging facilities all over the
+place, and they are mostly controlled with dedicated macros, that are
+implementation specific (to name a few, we have `GEN_CONV_DEBUG`,
+`TR_DEBUG`, `DEBUGPRINT`).
+
+Here we propose to unify those macros. The main benefit is, for
+developers, that the mechanism to print and debug would be the same
+across implementations.
+
+This macro should be independent of `NDEBUG` as we might need to use
+this logging information on the release build. We propose to call this
+unified macro `ONEDNN_DEVEL_MODE`, and it should have a matching cmake
+option.
+
+We recommend that all messages related to internal logic go under
+`ONEDNN_DEVEL_MODE`. Only messages related to user visible APIs which
+are actionable by user should live outside `ONEDNN_DEVEL_MODE`.
+
+## 2.3 Verbose levels vs verbose flags
+
+Currently, we have some documented information messages we provide
+(primitive creation/execution time), and some undocumented
+(e.g. runtimes info and implementation details).  For the undocumented
+ones, we don't have alignment on which level they should print (some
+print at level 2 and above, other print at level 5 and above).
+
+The current verbose levels were designed with the assumption that each
+level of information was inclusive of the previous ones. If we want to
+increase the number of categories of information to print
+(e.g. dispatch info, perf, error info, debug info, ...), it will
+quickly bloat the verbose output, which can become an issue when using
+verbose for long running workloads.  More importantly, verbose
+messages might impact runtime (e.g. getting perf on GPU), so getting
+some information sometimes impacts runtime.
+
+We propose here to use flags in order to limit verbose bloat, and
+allow printing messages without impacting runtime when possible. This
+could take two forms:
+- a list of message types, like `ERROR,PROFILE,DISPATCH,DEVINFO`...
```

Comment:
Another potentially useful feature is component/functionality filter like [--vmodule in glog](https://github.com/google/glog#verbose-logging). E.g. when setting `DNNL_VERBOSE=devinfo` with benchdnn we may want to avoid undesired output say from reorder (which is heavily used by benchdnn) or from CPU convolution (which is used as reference for GPU testing). We could support filters via something like `DNNL_VERBOSE=gpu_conv:dev_info,gpu_matmul:dev_info`.

Reply:
This is a good point. I will add that to the RFC